File: generated_code/cpp/solution_687_2.cpp
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int longestUnivaluePath(TreeNode* root) {
        int max_path = 0;
        dfs(root, max_path);
        return max_path;
    }

private:
    // This helper function returns the longest single-direction path starting 
    // from the current node going downwards, where all nodes have the same value.
    int dfs(TreeNode* node, int& max_path) {
        if (!node) return 0;

        // Recursively find the longest univalue path lengths in the left and right subtrees
        int left_len = dfs(node->left, max_path);
        int right_len = dfs(node->right, max_path);

        int left_arrow = 0;
        int right_arrow = 0;

        // If left child exists and has the same value, extend the left path
        if (node->left && node->left->val == node->val) {
            left_arrow = left_len + 1;
        }

        // If right child exists and has the same value, extend the right path
        if (node->right && node->right->val == node->val) {
            right_arrow = right_len + 1;
        }

        // Update the global maximum path which can combine both left and right directions
        max_path = std::max(max_path, left_arrow + right_arrow);

        // Return the max single-direction path extending from this node to its parent
        return std::max(left_arrow, right_arrow);
    }
};

File: generated_code/python/solution_687_2.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        """
        Algorithm: Post-order Traversal (DFS)
        
        For each node, we recursively find the longest univalue path extending into 
        its left and right subtrees. 
        - If the child exists and has the same value as the current node, we extend 
          the path length from that child by 1. Otherwise, the path from that direction is 0.
        - The longest univalue path *passing through* the current node as the highest 
          point is the sum of the left and right extended paths.
        - We maintain a global maximum (`max_path`) to track the longest path found.
        - The function returns the maximum single-direction path (left or right) to 
          allow the parent node to extend it.
          
        Time Complexity: O(N) where N is the number of nodes, as we visit each node once.
        Space Complexity: O(H) where H is the height of the tree, due to the recursion stack.
        """
        self.max_path = 0
        
        def dfs(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            
            # Recursively find the longest univalue paths in left and right subtrees
            left_len = dfs(node.left)
            right_len = dfs(node.right)
            
            # Variables to store the extended path from the current node
            left_arrow = right_arrow = 0
            
            # If left child exists and has the same value, extend the path
            if node.left and node.left.val == node.val:
                left_arrow = left_len + 1
                
            # If right child exists and has the same value, extend the path
            if node.right and node.right.val == node.val:
                right_arrow = right_len + 1
            
            # Update the global maximum with the combined path through the current node
            self.max_path = max(self.max_path, left_arrow + right_arrow)
            
            # Return the longest single-direction path to the parent
            return max(left_arrow, right_arrow)
        
        dfs(root)
        return self.max_path